### PR TITLE
Ebs creation script

### PR DESCRIPTION
**Pull request by MattPinelli**
Script and modules to create an EBS snapshot with the contents of a given LMS bundle.

### CODE REVIEW

File: scripts/lib/EC2Helpers.psm1 [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+        -File "C:\$PackageVersion.zip"
+}
+
+function Expand-LMSPackage {
+    param (
+        [Parameter(Mandatory = $True)][string] $PackageVersion,
+        [Parameter(Mandatory = $True)][string] $DriveLetter
+    )
+    Invoke-Expression "C:\PROGRA~1\7-Zip\7z.exe x C:\$PackageVersion.zip -o$DriveLetter"
+    Start-Sleep 5
```
Comment (jtgrohn) [inline scripts/lib/EC2Helpers.psm1]: This sleep is weird. Why do we need it? Is 5 seconds always going to be enough?

File: scripts/lib/EC2Helpers.psm1 [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+}
+
+function New-TagSpec {
+    param (
+        [Parameter(Mandatory = $True)][string] $PackageVersion
+    )
+    $tag = @{ Key = "Name"; Value = $PackageVersion }
```
Comment (jtgrohn) [inline scripts/lib/EC2Helpers.psm1]: :nit I generally prefer constucting hashtables over multiple lines, but for two properties, this is fine.
```
$tag = @{
     Key = "Name"
     Value = $PackageVersion
}
```

File: scripts/lib/EC2Helpers.psm1 [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+    }
+}
+
+function Dismount-LMSVolume {
+    param (
+        [Parameter(Mandatory = $True)][string] $InstanceId,
+        [Parameter(Mandatory = $True)][string] $VolumeId
+    )
+    Dismount-EC2Volume -InstanceId $InstanceId `    
+    -VolumeId $VolumeId
```
Comment (jtgrohn) [inline scripts/lib/EC2Helpers.psm1]: :nit, tab this line in

File: scripts/lib/EC2Helpers.psm1 [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+    param (
+        [Parameter(Mandatory = $True)][string] $VolumeId
+    )
+    Remove-EC2Volume -VolumeId $VolumeId `
+        -Force
+}
+
+function Get-AvailableDeviceName {
+    param(            
+        [parameter (mandatory = $true)][string] $InstanceId
```
Comment (jtgrohn) [inline scripts/lib/EC2Helpers.psm1]: :nit capitalize `Parameter` and `Mandatory`. Get rid of space after `parameter`

File: scripts/lib/DiskHelpers.psm1 [hunk 5]
```diff
@@ -0,0 +1,14 @@
+function Get-AvailableDriveLetter {
+    $latestDriveLetter = Get-PSDrive `
+        -PSProvider FileSystem `
+    | Select-Object -ExpandProperty Name `
```
Comment (jtgrohn) [inline scripts/lib/DiskHelpers.psm1]: :nit, tab this line in.
Also, you can do this with only one pipeline. 
```
$latestDriveLetter = Get-PSDrive `
     -PSProvider FileSystem `
     | Select-Object -ExpandProperty Name -Last 1
```

File: scripts/lib/EC2Helpers.psm1 [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+        [parameter (mandatory = $true)][string] $InstanceId            
+    )            
+    $currentDevice = Get-EC2InstanceAttribute $InstanceId `
+        -Attribute blockDeviceMapping `
+    | Select-Object -ExpandProperty BlockDeviceMappings `
+    | Select-Object -last 1
+    
+    if ($currentDevice.DeviceName -eq '/dev/sda1') {            
+        return "xvdf"
+    }
```
Comment (jtgrohn) [inline scripts/lib/EC2Helpers.psm1]: noticing some code styles you do differently than everybody else.
Generally folks at D2L (when writing .NET code) go for spaces inside parentheses 
```
if( $currentDevice.DeviceName -eq '/dev/sda1' ) {
```
and no new lines before open squigglies
``` 
} else {
```

Comment (jtgrohn) [reply scripts/lib/EC2Helpers.psm1]: :nit. Could remove one of the clauses of this `if`-`else` by having a default value and negating the `if` condition
```
$deviceName = "xvdf"
if( $currentDevice.DeviceName -ne '/dev/sda1' ) {
     $device = $currentDevice.DeviceName.ToCharArray()
     $device[3] = [char]( [int]( $device[3] ) + 1 )
     $deviceName  = -join $device
     Write-Host "Next available Device Name is $deviceName"
}
return $deviceName
```

File: scripts/lib/LockHelpers.psm1 [hunk 7]
```diff
@@ -0,0 +1,21 @@
+function New-Lock {
```
Comment (jtgrohn) [inline scripts/lib/LockHelpers.psm1]: I don't think this works 100% of the time. Will get back to you on it.

File: scripts/lib/EC2Helpers.psm1 (hunk 1 above)
Comment (MattPinelli) [reply scripts/lib/EC2Helpers.psm1]: I found that for whatever reason, the snapshots would be missing files if there was no sleep.  I have found 5 seconds to be fine.

File: scripts/lib/LockHelpers.psm1 (hunk 7 above)
Comment (MattPinelli) [reply scripts/lib/LockHelpers.psm1]: OK sounds good.

File: scripts/lib/EC2Helpers.psm1 (hunk 1 above)
Comment (jtgrohn) [reply scripts/lib/EC2Helpers.psm1]: Is there a test we can do instead?

File: scripts/lib/EC2Helpers.psm1 [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+        -Key "$PackageVersion.zip" `
+        -File "C:\$PackageVersion.zip"
+}
+
+function Expand-LMSPackage {
+    param (
+        [Parameter(Mandatory = $True)][string] $PackageVersion,
+        [Parameter(Mandatory = $True)][string] $DriveLetter
+    )
+    Invoke-Expression "C:\PROGRA~1\7-Zip\7z.exe x C:\$PackageVersion.zip -o$DriveLetter"
```
[the remark this answers is not in the document]
Comment (MattPinelli) [reply scripts/lib/EC2Helpers.psm1]: As we discussed, 7zip is much faster than Expand-Archive. I will check the exit code before continuing.

**Pull request merged**